src/libcrun/handlers/wamr.c

```diff
@@ -0,0 +1,249 @@
+/*
+ * crun - OCI runtime written in C
+ *
+ * Copyright (C) 2017, 2018, 2019, 2020, 2021 Giuseppe Scrivano <giuseppe@scrivano.org>
+ * crun is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * crun is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with crun.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#define _GNU_SOURCE
+
+#include <config.h>
+#include "../custom-handler.h"
+#include "../container.h"
+#include "../utils.h"
+#include "../linux.h"
+#include "handler-utils.h"
+#include <unistd.h>
+#include <sys/stat.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <sched.h>
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#ifdef HAVE_DLOPEN
+#  include <dlfcn.h>
+#endif
+
+#ifdef HAVE_WAMR
+#  include <wasm_export.h>
+#endif
+
+#if HAVE_DLOPEN && HAVE_WAMR
+
+static int
+libwamr_load (void **cookie, libcrun_error_t *err)
+{
+  void *handle;
+
+  handle = dlopen ("libiwasm.so", RTLD_NOW);
+  if (handle == NULL)
+    return crun_make_error (err, 0, "could not load `libiwasm.so`: `%s`", dlerror ());
+  *cookie = handle;
+
+  return 0;
+}
+
+static int
+libwamr_unload (void *cookie, libcrun_error_t *err)
+{
+  int r;
+
+  if (cookie)
+    {
+      r = dlclose (cookie);
+      if (UNLIKELY (r < 0))
+        return crun_make_error (err, 0, "could not unload handle: `%s`", dlerror ());
+    }
+  return 0;
+}
+
+// Function to read a WebAssembly binary file into a buffer
+char *read_wasm_binary_to_buffer(const char *pathname, uint32_t *size) {
+
+    FILE *file;
+    char *buffer;
+    size_t file_size;
+
+    // Open the file in binary mode
+    file = fopen(pathname, "rb");
+    if (!file) {
+        error (EXIT_FAILURE, 0, "Failed to open file");
+        return NULL;
+    }
+
+    // Seek to the end of the file to determine the size
+    fseek(file, 0, SEEK_END);
+    file_size = ftell(file);
+    fseek(file, 0, SEEK_SET);
+
+    // Allocate memory for the buffer
+    buffer = (char *)malloc(file_size);
```

Comment:
crun uses `xmalloc` across the code base, i think it should exit on its own when failure happens.

Reply:
@giuseppe WDYT ?